### ISSUE TITLE
"Could not load type" for nested generated enum.

### ISSUE THREAD

**Issue by takemyoxygen** [fsprojects/FSharp.TypeProviders.StarterPack#102]
### Description

After #101 was merged, I moved one step closer to supporting [enums in protobuf type provider](https://github.com/takemyoxygen/ProtoTypes/tree/cli-enums) and discovered another issue: when enum is a nested type and enclosing type has a field of this nested enum type, then provider throws an exception `Could not load type 'MyEnum' from assembly 'tmp193C, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.` when `TypeBuilder.CreateType()` is called [here] (https://github.com/fsprojects/FSharp.TypeProviders.StarterPack/blob/master/src/ProvidedTypes.fs#L2741).

### Repro
```
[<TypeProvider>]
type EnumProvider() as this = 
    inherit TypeProviderForNamespaces()
    
    let ns = "EnumTypeProvider.Provided"
    let asm = Assembly.GetExecutingAssembly()
    let tempAsm = Path.ChangeExtension(Path.GetTempFileName(), ".dll") |> ProvidedAssembly
    let container = ProvidedTypeDefinition(asm, ns, "Container", Some typeof<obj>, IsErased = false)
    
    do
        // Generate a class
        let ty = ProvidedTypeDefinition("MyClass", Some typeof<obj>, IsErased = false)
        ty.AddMember <| ProvidedConstructor([], InvokeCode = (fun _ -> Expr.Value(())))

        // Generate an enum
        let enum = ProvidedTypeDefinition("MyEnum", Some typeof<Enum>, IsErased = false)
        enum.SetEnumUnderlyingType typeof<int>
        enum.AddMember <| ProvidedLiteralField("Foo", enum, 1)
        enum.AddMember <| ProvidedLiteralField("Bar", enum, 2)
        ty.AddMember enum
         
        // Generate a field and a property
        let field = ProvidedField("myField", enum)
        ty.AddMember field
        
        ty.AddMember <| ProvidedProperty(
            "MyProperty",
            enum,
            GetterCode = (fun (this::_) -> Expr.FieldGet(this, field)),
            SetterCode = (fun (this::value::_) -> Expr.FieldSet(this, field, value)))
        container.AddMember ty

        tempAsm.AddTypes [container]
        this.AddNamespace(ns, [container])
        
[<assembly:TypeProviderAssembly>] 
do()
```

If this type provided is used like this:
```
type Container = EnumTypeProvider.Provided.Container
let cs = Container.MyClass()
cs.MyProperty <- Container.MyClass.MyEnum.Bar
```

I get the following error:
```
Script.fsx(4,18): error FS3033: The type provider 'EnumTypeProvider.EnumProvider' reported an error: 
Could not load type 'MyEnum' from assembly 'tmp1694, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
```
### Details

The error happens only on Windows. On Mono 4.3.0 it works as expected. I think this is exactly the same situation described in [the `TypeBuilder.CreateType()` docs](https://msdn.microsoft.com/en-us/library/system.reflection.emit.typebuilder.createtype(v=vs.110).aspx#Anchor_2).

I was able to fix it by adding the following code to `ProvidedTypes.fs` instead of [this line](https://github.com/fsprojects/FSharp.TypeProviders.StarterPack/blob/master/src/ProvidedTypes.fs#L2741):
```
let completed = HashSet<_>()

let complete (tb: TypeBuilder) =
    if not (completed.Contains tb) then 
        tb.CreateType() |> ignore
        completed.Add tb |> ignore

let domain = System.Threading.Thread.GetDomain()

let resolveHandler = ResolveEventHandler(fun _ args ->
    typeMap.Values
    |> Seq.filter (fun tb -> tb.Name = args.Name)
    |> Seq.iter complete
    assemblyMainModule.Assembly)

domain.add_TypeResolve resolveHandler

iterateTypes (fun tb _ptd -> complete tb)

domain.remove_TypeResolve resolveHandler
```

I'll probably have a look at what downsides this fix might have (e.g. nested classes with same names etc) and submit a PR.
